release-23.1: server/license: store cluster init time in the KV system keyspace

spilchen opened the pull request:
Backport 1/1 commits from #129634.

/cc @cockroachdb/release

---

We need to store cluster-level metadata for license policy enforcement. This metadata must be protected from modification, so it cannot be stored in any system table. This PR introduces a new field in the system keyspace to store the cluster's initialization start time, which will be used in a follow-up PR to track the grace period for clusters without a license. The metadata needs to be stored at the cluster level to align with the license's scope.

A new struct, License.Enforcer, is responsible for writing and caching the new timestamp. This struct will be extended in future PRs to check for throttling in clusters that violate our license policies.

This update will be backported to versions 23.1, 23.2, 24.1, and 24.2.

Epic: CRDB-39988
Informs: CRDB-40064
Release note: None
Release justification: This work is part of the CockroachDB core deprecation.

spilchen committed:
server/license: store cluster init time in the KV system keyspace

We need to store cluster-level metadata for license policy enforcement.
This metadata must be protected from modification, so it cannot be
stored in any system table. This PR introduces a new field in the system
keyspace to store the cluster's initialization start time, which will be
used in a follow-up PR to track the grace period for clusters without a
license. The metadata needs to be stored at the cluster level to align
with the license's scope.

A new struct, License.Enforcer, is responsible for writing and caching
the new timestamp. This struct will be extended in future PRs to check
for throttling in clusters that violate our license policies.

This update will be backported to versions 23.1, 23.2, 24.1, and 24.2.

Epic: CRDB-39988
Informs: CRDB-40064
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/129806)
<!-- Reviewable:end -->

rafiss reviewed (commented):
looks good, but we should avoid merging this until we are close to cutting the October release for 23.1. otherwise, the grace period will start too early for clusters that upgrade to a new patch version before October.

spilchen committed:
server/license: disable write of grace period init timestamp

This follows up on commit 456521897e0, where we introduced a new grace
period initialization timestamp written to the system key space for
license enforcement. We were concerned that if we backported this change
and released a patch update before the epic is done, customers might
start the clock for the grace period without having the full license
deprecation code in place. This could lead to immediate throttling once
they do upgraded to a release with the complete code. To mitigate this,
we are making the writing of the grace period initialization timestamp
opt-in and limited to testing scenarios. We will enable it by default
once more of the core deprecation work is completed.

This change will be backported to 23.1, 23.2, 24.1 and 24.2.

Epic: CRDB-39988
Informs: CRDB-40064
Release note: None

spilchen commented:
This includes the backport for #129898 as well. See the discussion in #129791. Both changes need to be merged together to ensure that no part of the core license deprecation is enabled until the entire implementation is complete.

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 13 of 13 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129806#-:-O5xZvFRDRHbbbB-m3Hu:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2)



<!-- Sent from Reviewable.io -->
